etcdserver/cindex/cindex.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2015 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cindex
+
+import (
+	"encoding/binary"
+	"go.etcd.io/etcd/mvcc/backend"
+	"sync/atomic"
+)
+
+var (
+	metaBucketName = []byte("meta")
+
+	consistentIndexKeyName = []byte("consistent_index")
+)
+
+// ConsistentIndexer is an interface that wraps the Get/Save method.
+// Consistent index is the offset of an entry in a consistent replicated log.
+type ConsistentIndexer interface {
+	// ConsistentIndex returns the consistent index of current executing entry.
+	ConsistentIndex() uint64
+
+	// Save saves consistentIndex to the underlying stable storage.
+	Save(tx backend.BatchTx)
+}
+
+// consistentIndex represents the offset of an entry in a consistent replica log.
+// It implements the ConsistentIndexer interface.
+// It is always set to the offset of current entry before executing the entry,
+// so ConsistentWatchableKV could get the consistent index from it.
+
+type ConsistentIndex struct {
+	consistentIndex uint64
+	bytesBuf8       []byte
+}
+
+func NewConsistentIndex() ConsistentIndex {
+	return ConsistentIndex{bytesBuf8: make([]byte, 8)}
+}
+
+func (cindex *ConsistentIndex) LoadConsistentIndex(tx backend.BatchTx) uint64 {
```

Comment:
Is there a particular reason for having a separate function to load index? In the old implementation this is part of ConsistentIndex().

Reply:
thanks. if we do not have LoadConsistentIndex method, we have to pass batch.Tx parameter to ConsistentIndex Method(consistent index is zero after starting etcd server).

Reply:
Have you tried to pass batch Tx when creating new consistent index? Do you feel it provides better encapsulation?

Reply:
i have tried to pass batch Tx when creating new consistent index. there are so many test case parameters need to be changed and it seems  batch Tx be closed sometimes. for example, defrag will set batch.tx to nil.  I will deep dive it tomorrow.

```
anic: assertion failed: tx closed
goroutine 895 [running]:
go.etcd.io/etcd/vendor/go.etcd.io/bbolt._assert(...)
 /Users/simotang/go/src/go.etcd.io/etcd/vendor/go.etcd.io/bbolt/db.go:1172
go.etcd.io/etcd/vendor/go.etcd.io/bbolt.(*Cursor).seek(0xc006f4ede0, 0x2556520, 0x4, 0x4, 0x1bfeea9, 0x1d, 0x11, 0x2, 0x0, 0x0, ...)
 /Users/simotang/go/src/go.etcd.io/etcd/vendor/go.etcd.io/bbolt/cursor.go:155 +0x182
go.etcd.io/etcd/vendor/go.etcd.io/bbolt.(*Bucket).Bucket(0xc0001c8718, 0x2556520, 0x4, 0x4, 0x0)
 /Users/simotang/go/src/go.etcd.io/etcd/vendor/go.etcd.io/bbolt/bucket.go:105 +0xd5
```

Reply:
i see. oldbe will be closed when etcdserver execute applySnapshot. i try to add a SetBatchTx method to solve this issue. how do you think so ?
```
	lg.Info("restored mvcc store")

	// Closing old backend might block until all the txns
	// on the backend are finished.
	// We do not want to wait on closing the old backend.
	s.bemu.Lock()
	oldbe := s.be
	go func() {
		lg.Info("closing old backend file")
		defer func() {
			lg.Info("closed old backend file")
		}()
		if err := oldbe.Close(); err != nil {
			lg.Panic("failed to close old backend", zap.Error(err))
		}
	}()

	s.be = newbe
	s.bemu.Unlock()
```

Reply:
ConsistentIndex and the store that uses it needs to make sure they always refer to the same backend. The store can actually close backend and open a new backend, whenever this happens, it needs to update the backend for consistentIndex. Is my understand correct? 

If this is the case, I don't feel strongly which implementation has better encapsulation. Maybe keep the old way? (i.e., the caller of Load and Save methods needs to pass in the batchTx of the backend it is currently using)

Reply:
yes,we have to update  the backend used by consistentIndex when old backend be closed. no other commands will close backend except applySnapshot. i think it is safe.  if we keep the old way, we also have to call LoadConsistentIndex when backend being closed.  meanwhile, we must firstly call LoadConsistentIndex  when creating a NewConsistentIndex every time in old version. By comparison, the new version is more friendly.

Reply:
if we call SetBatchTx in store.Restore function, it seems very good to solve this issue. there are only two function(NewStore/Restore) will update backend.
```
--- a/mvcc/kvstore.go
+++ b/mvcc/kvstore.go
@@ -333,6 +333,8 @@ func (s *store) Restore(b backend.Backend) error {
        s.compactMainRev = -1
        s.fifoSched = schedule.NewFIFOScheduler()
        s.stopc = make(chan struct{})
+       s.ci.SetConsistentIndex(0)
+       s.ci.SetBatchTx(b.BatchTx())
```

Reply:
OK let's go with the new implementation.

Reply:
thanks. i have updated pr and travis ci passes all test cases.